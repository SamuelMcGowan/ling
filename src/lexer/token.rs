use serde::Serialize;

use crate::constants::ConstIdx;
use crate::syntax::source::Span;

#[derive(Debug, Clone, Copy, Serialize)]
pub(crate) struct Token {
    pub span: Span,
    pub kind: TokenKind,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub(crate) enum TokenKind {
    Punct(Punct),
    Bracket(Bracket),

    Keyword(Keyword),

    Ident(Ustr),
    Const(ConstIdx),

    Error(Error),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub(crate) enum Punct {
    Dot,
    Comma,
    Colon,
    Semicolon,

    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Pow,

    AddEqual,
    SubEqual,
    MulEqual,
    DivEqual,
    ModEqual,
    PowEqual,

    Gt,
    Lt,
    Equal,

    GtEqual,
    LtEqual,
    EqualEqual,

    BangEqual,
    Bang,

    FatArrow,
    Arrow,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub(crate) enum Bracket {
    Opening(BracketKind),
    Closing(BracketKind),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub(crate) enum BracketKind {
    Round,
    Curly,
    Square,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub(crate) enum Keyword {
    Func,
    Return,

    Type,
    Enum,
    Data,
    Typeclass,
    Impl,

    Let,

    Match,
    Is,

    LogicalAnd,
    LogicalOr,
    LogicalNot,

    If,
    Elif,
    Else,

    Loop,
    While,

    Unit,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub(crate) enum Error {
    UnexpectedChar(char),

    InvalidEscape,
    UnterminatedString,

    IntegerOverflow,
    MissingDigits,
}

macro_rules! tkind {
    (punct $punct:ident) => {{
        use crate::lexer::token::*;
        TokenKind::Punct(Punct::$punct)
    }};
    (bracket $side:ident $bracket:ident) => {{
        use crate::lexer::token::*;
        TokenKind::Bracket(Bracket::$side(BracketKind::$bracket))
    }};

    (kwd $kwd:ident) => {{
        use crate::lexer::token::*;
        TokenKind::Keyword(Keyword::$kwd)
    }};

    (ident $ident:expr) => {{
        use ustr::Ustr;
        use crate::lexer::token::*;

        TokenKind::Ident(Ustr::from($ident))
    }};
    (constant $constant:literal) => {{
        use crate::lexer::token::*;
        use crate::constants::ConstIdx;

        TokenKind::Const(ConstIdx($constant))
    }};

    (error $error:ident $($value:tt)?) => {{
        use crate::lexer::token::*;
        TokenKind::Error(Error::$error $($value)?)
    }}
}

pub(crate) use tkind;
use ustr::Ustr;
