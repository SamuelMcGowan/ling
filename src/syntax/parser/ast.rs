use macro_rules_attribute::{derive, derive_alias};
use ustr::Ustr;

use crate::constants::ConstIdx;

derive_alias! {
    #[derive(Node!)] = #[derive(
        Debug,
        Clone,
        serde::Serialize,
        serde::Deserialize,
    )];
}

#[derive(Node!)]
pub(crate) struct Module {
    pub items: Vec<Item>,
}

#[derive(Node!)]
pub(crate) enum Item {
    Func(Func),
}

#[derive(Node!)]
pub(crate) struct Func {
    pub ident: Ident,
    pub params: Vec<(Ident, Ty)>,
    pub ret_ty: Ty,
    pub body: Expr,
}

#[derive(Node!)]
pub(crate) enum Ty {
    Constructed { ident: Ident, params: Vec<Ty> },
    Unit,
}

#[derive(Node!)]
pub(crate) enum Stmt {
    Expr(Expr),
}

#[derive(Node!)]
pub(crate) enum Expr {
    Const(ConstIdx),
    Unit,
}

#[derive(Node!)]
pub(crate) enum Ident {
    Unresolved(Ustr),
}

impl Ident {
    pub fn unresolved(&self) -> Option<Ustr> {
        match self {
            Self::Unresolved(s) => Some(*s),
        }
    }
}
