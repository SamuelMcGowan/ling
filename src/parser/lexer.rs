use super::span::Span;
use super::token::{tkind, Token};
use super::ParseContext;

impl<'a> ParseContext<'a> {
    pub fn lex_token(&mut self) -> Option<Token> {
        loop {
            let c = self.cursor.next()?;
            let start_pos = self.cursor.byte_pos();

            let kind = match c {
                c if c.is_ascii_whitespace() => continue,
                '#' => {
                    while !matches!(self.cursor.next(), Some('\n') | None) {}
                    continue;
                }

                '(' => tkind!(punct LParen),
                ')' => tkind!(punct RParen),
                '{' => tkind!(punct LBrace),
                '}' => tkind!(punct RBrace),
                '[' => tkind!(punct LBracket),
                ']' => tkind!(punct RBracket),

                '.' => tkind!(punct Dot),
                ',' => tkind!(punct Comma),
                ':' => tkind!(punct Colon),
                ';' => tkind!(punct Semicolon),

                '-' if self.cursor.eat('>') => tkind!(punct Arrow),
                '=' if self.cursor.eat('>') => tkind!(punct FatArrow),

                '+' if self.cursor.eat('=') => tkind!(punct AddEqual),
                '-' if self.cursor.eat('=') => tkind!(punct SubEqual),
                '*' if self.cursor.eat('=') => tkind!(punct MulEqual),
                '/' if self.cursor.eat('=') => tkind!(punct DivEqual),
                '%' if self.cursor.eat('=') => tkind!(punct ModEqual),
                '^' if self.cursor.eat('=') => tkind!(punct PowEqual),

                '+' => tkind!(punct Add),
                '-' => tkind!(punct Sub),
                '*' => tkind!(punct Mul),
                '/' => tkind!(punct Div),
                '%' => tkind!(punct Mod),
                '^' => tkind!(punct Pow),

                '>' if self.cursor.eat('=') => tkind!(punct GtEqual),
                '<' if self.cursor.eat('=') => tkind!(punct LtEqual),
                '=' if self.cursor.eat('=') => tkind!(punct EqualEqual),

                '>' => tkind!(punct Gt),
                '<' => tkind!(punct Lt),
                '=' => tkind!(punct Equal),

                _ => tkind!(error UnexpectedChar(c)),
            };

            let span = Span::new(start_pos, self.cursor.byte_pos());

            return Some(Token { span, kind });
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::token::TokenKind;

    #[test]
    fn punct_single() {
        check_tokens(
            "+-:;",
            &[
                tkind!(punct Add),
                tkind!(punct Sub),
                tkind!(punct Colon),
                tkind!(punct Semicolon),
            ],
        );
    }

    #[test]
    fn punct_double() {
        check_tokens(
            "->==+=",
            &[
                tkind!(punct Arrow),
                tkind!(punct EqualEqual),
                tkind!(punct AddEqual),
            ],
        )
    }

    #[test]
    fn whitespace() {
        check_tokens(" .\r\n. ", &[tkind!(punct Dot), tkind!(punct Dot)])
    }

    #[test]
    fn comment() {
        check_tokens(".# comment\r\n.# comment", &[tkind!(punct Dot), tkind!(punct Dot)]);
    }

    #[test]
    fn comment_empty() {
        check_tokens(".#\r\n.#", &[tkind!(punct Dot), tkind!(punct Dot)])
    }

    fn check_tokens(s: &str, t: &[TokenKind]) {
        let mut context = ParseContext::new(s);

        let mut tokens = vec![];
        while let Some(token) = context.lex_token() {
            tokens.push(token.kind);
        }

        assert_eq!(&tokens, t)
    }
}
