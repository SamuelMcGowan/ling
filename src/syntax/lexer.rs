use ustr::Ustr;

use super::span::Span;
use super::token::{tkind, Token, TokenKind};
use super::{Cursor, ParseContext};
use crate::value::Value;

impl<'a> ParseContext<'a> {
    pub fn lex_tokens(&mut self) -> Lexer<'_, 'a> {
        Lexer {
            cursor: Cursor::new(self.source),
            context: self,
        }
    }
}

pub(crate) struct Lexer<'ctx, 'a> {
    cursor: Cursor<'a>,
    context: &'ctx mut ParseContext<'a>,
}

impl Lexer<'_, '_> {
    fn lex_token(&mut self) -> Option<Token> {
        loop {
            let start_pos = self.cursor.byte_pos();
            let c = self.cursor.next()?;

            let kind = match c {
                c if c.is_ascii_whitespace() => continue,
                '#' => {
                    while !matches!(self.cursor.next(), Some('\n') | None) {}
                    continue;
                }

                '(' => tkind!(punct LParen),
                ')' => tkind!(punct RParen),
                '{' => tkind!(punct LBrace),
                '}' => tkind!(punct RBrace),
                '[' => tkind!(punct LBracket),
                ']' => tkind!(punct RBracket),

                '.' => tkind!(punct Dot),
                ',' => tkind!(punct Comma),
                ':' => tkind!(punct Colon),
                ';' => tkind!(punct Semicolon),

                '-' if self.cursor.eat('>') => tkind!(punct Arrow),
                '=' if self.cursor.eat('>') => tkind!(punct FatArrow),

                '+' if self.cursor.eat('=') => tkind!(punct AddEqual),
                '-' if self.cursor.eat('=') => tkind!(punct SubEqual),
                '*' if self.cursor.eat('=') => tkind!(punct MulEqual),
                '/' if self.cursor.eat('=') => tkind!(punct DivEqual),
                '%' if self.cursor.eat('=') => tkind!(punct ModEqual),
                '^' if self.cursor.eat('=') => tkind!(punct PowEqual),

                '+' => tkind!(punct Add),
                '-' => tkind!(punct Sub),
                '*' => tkind!(punct Mul),
                '/' => tkind!(punct Div),
                '%' => tkind!(punct Mod),
                '^' => tkind!(punct Pow),

                '>' if self.cursor.eat('=') => tkind!(punct GtEqual),
                '<' if self.cursor.eat('=') => tkind!(punct LtEqual),
                '=' if self.cursor.eat('=') => tkind!(punct EqualEqual),

                '>' => tkind!(punct Gt),
                '<' => tkind!(punct Lt),
                '=' => tkind!(punct Equal),

                _ if is_ident_start(c) => self.lex_ident(start_pos),
                '0'..='9' => self.lex_number(c as u8),
                '"' => self.lex_string(),

                _ => tkind!(error UnexpectedChar(c)),
            };

            let span = Span::new(start_pos, self.cursor.byte_pos());

            return Some(Token { span, kind });
        }
    }

    fn lex_ident(&mut self, start_pos: usize) -> TokenKind {
        while matches!(self.cursor.peek(), Some(c) if is_ident(c)) {
            self.cursor.next();
        }

        let lexeme = &self.cursor.as_str_all()[start_pos..self.cursor.byte_pos()];

        match lexeme {
            "func" => tkind!(kwd Func),
            "return" => tkind!(kwd Return),

            "type" => tkind!(kwd Type),
            "enum" => tkind!(kwd Enum),
            "data" => tkind!(kwd Data),
            "typeclass" => tkind!(kwd Typeclass),
            "impl" => tkind!(kwd Impl),

            "let" => tkind!(kwd Let),
            "match" => tkind!(kwd Match),

            "if" => tkind!(kwd If),
            "elif" => tkind!(kwd Elif),
            "else" => tkind!(kwd Else),

            "loop" => tkind!(kwd Loop),
            "while" => tkind!(kwd While),

            "true" => self.add_constant(Value::Bool(true)),
            "false" => self.add_constant(Value::Bool(false)),

            _ => tkind!(ident lexeme),
        }
    }

    fn lex_string(&mut self) -> TokenKind {
        let mut s = String::new();

        let mut unterminated_string = true;
        let mut invalid_escape = false;

        while let Some(c) = self.cursor.next() {
            let c = match c {
                '"' => {
                    unterminated_string = false;
                    break;
                }
                '\\' => match self.cursor.next() {
                    Some(c) => match c {
                        'n' => '\n',
                        'r' => '\n',
                        't' => '\t',
                        '0' => '\0',
                        _ => {
                            invalid_escape = true;
                            c
                        }
                    },
                    None => return tkind!(error UnterminatedString),
                },
                _ => c,
            };
            s.push(c);
        }

        if unterminated_string {
            return tkind!(error UnterminatedString);
        }
        if invalid_escape {
            return tkind!(error InvalidEscape);
        }

        let interned = Ustr::from(&s);
        self.add_constant(Value::String(interned))
    }

    fn lex_number(&mut self, first: u8) -> TokenKind {
        let int_digits = self.lex_digits(Some(first));

        if self.cursor.eat('.') {
            let fraction_digits = self.lex_digits(None);

            let mut exponent_overflow = false;
            let mut exponent_missing_digits = false;

            let exponent: i32 = if self.cursor.eat('e') || self.cursor.eat('E') {
                let sign = if self.cursor.eat('-') {
                    -1
                } else {
                    self.cursor.eat('+');
                    1
                };

                let exponent_digits = self.lex_digits(None);

                let exponent = parse_int(&exponent_digits).and_then(|n| n.try_into().ok());
                let exponent: i32 = match exponent {
                    Some(exponent) => exponent,
                    None => {
                        exponent_overflow = true;
                        0
                    }
                };

                exponent_missing_digits = exponent_digits.is_empty();

                exponent * sign
            } else {
                0
            };

            if exponent_overflow {
                return tkind!(error IntegerOverflow);
            }

            // integer won't have missing digits
            if fraction_digits.is_empty() || exponent_missing_digits {
                return tkind!(error MissingDigits);
            }

            let int_digits = trim_leading_zeros(&int_digits).iter();
            let fraction_digits = trim_trailing_zeros(&fraction_digits).iter();

            let f: f64 = minimal_lexical::parse_float(int_digits, fraction_digits, exponent);
            self.add_constant(Value::Float(f))
        } else {
            let Some(n) = parse_int(&int_digits) else {
                return tkind!(error IntegerOverflow);
            };
            self.add_constant(Value::Integer(n))
        }
    }

    fn lex_digits(&mut self, first: Option<u8>) -> Vec<u8> {
        let mut digits = vec![];
        digits.extend(first);
        loop {
            match self.cursor.peek() {
                Some(c @ '0'..='9') => {
                    digits.push(c as u8);
                    self.cursor.next();
                }
                Some('_') => continue,
                _ => break,
            }
        }
        digits
    }

    fn add_constant(&mut self, value: Value) -> TokenKind {
        let idx = self.context.chunk.add_constant(value);
        TokenKind::Const(idx)
    }
}

impl Iterator for Lexer<'_, '_> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        self.lex_token()
    }
}

fn is_ident_start(c: char) -> bool {
    c.is_ascii_alphabetic() || c == '_'
}

fn is_ident(c: char) -> bool {
    c.is_ascii_alphanumeric() || c == '_'
}

fn parse_int(digits: &[u8]) -> Option<u64> {
    let mut n = 0u64;
    for &digit in digits {
        let digit = (digit - b'0') as u64;
        n = n.checked_mul(10)?;
        n = n.checked_add(digit)?;
    }
    Some(n)
}

fn trim_leading_zeros(mut digits: &[u8]) -> &[u8] {
    while let [b'0', rest @ ..] = digits {
        digits = rest;
    }
    digits
}

fn trim_trailing_zeros(mut digits: &[u8]) -> &[u8] {
    while let [rest @ .., b'0'] = digits {
        digits = rest;
    }
    digits
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::chunk::{Chunk, ConstIdx};

    #[test]
    fn punct_single() {
        check_tokens(
            "+-:;",
            &[
                tkind!(punct Add),
                tkind!(punct Sub),
                tkind!(punct Colon),
                tkind!(punct Semicolon),
            ],
        );
    }

    #[test]
    fn punct_double() {
        check_tokens(
            "->==+=",
            &[
                tkind!(punct Arrow),
                tkind!(punct EqualEqual),
                tkind!(punct AddEqual),
            ],
        );
    }

    #[test]
    fn whitespace() {
        check_tokens(" .\r\n. ", &[tkind!(punct Dot), tkind!(punct Dot)]);
    }

    #[test]
    fn comment() {
        check_tokens(
            ".# comment\r\n.# comment",
            &[tkind!(punct Dot), tkind!(punct Dot)],
        );
    }

    #[test]
    fn comment_empty() {
        check_tokens(".#\r\n.#", &[tkind!(punct Dot), tkind!(punct Dot)]);
    }

    #[test]
    fn keyword() {
        check_tokens("else", &[tkind!(kwd Else)]);
    }

    #[test]
    fn ident() {
        check_tokens("else_", &[tkind!(ident "else_")])
    }

    #[test]
    fn literals() {
        check_constant("true", Value::Bool(true));
        check_constant("false", Value::Bool(false));
    }

    #[test]
    fn string() {
        check_constant("\"hello\"", Value::String(Ustr::from("hello")));
        check_constant("\"\"", Value::String(Ustr::from("")));
    }

    #[test]
    fn string_unterminated() {
        check_tokens("\"hello", &[tkind!(error UnterminatedString)]);
        check_tokens("\"", &[tkind!(error UnterminatedString)]);
    }

    #[test]
    fn integer() {
        check_constant("123", Value::Integer(123));
    }

    #[test]
    fn float() {
        check_constant("12.34e-5", Value::Float(12.34e-5));
    }

    #[test]
    fn missing_digits() {
        check_tokens("12.", &[tkind!(error MissingDigits)]);
        check_tokens("12.1e", &[tkind!(error MissingDigits)]);
    }

    fn check_tokens(s: &str, t: &[TokenKind]) {
        let mut chunk = Chunk::default();
        let mut context = ParseContext::new(s, &mut chunk);
        let tokens: Vec<_> = context.lex_tokens().map(|token| token.kind).collect();
        assert_eq!(&tokens, t)
    }

    fn check_constant(s: &str, value: Value) {
        let mut chunk = Chunk::default();
        let mut context = ParseContext::new(s, &mut chunk);

        let tokens: Vec<_> = context.lex_tokens().map(|token| token.kind).collect();
        assert_eq!(&tokens, &[tkind!(constant 0)]);

        let value_found = chunk.get_constant(ConstIdx(0)).expect("constant not found");
        assert!(value_found.is_equal(&value));
    }
}
