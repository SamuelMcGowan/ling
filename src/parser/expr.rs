use super::{ParseError, ParseResult, Parser};
use crate::ast::*;
use crate::lexer::token::{tkind, BracketKind, TokenKind};
use crate::syntax::token_stream::TokenTree;

#[repr(u8)]
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
enum Prec {
    Lowest = 0,

    LogicalOr,
    LogicalAnd,

    Equality,
    Comparison,

    Term,
    Factor,
    Exp,

    Unary,

    Call,
    Access,
}

impl BinOp {
    fn should_parse_in_prec(&self, prec: Prec) -> bool {
        let op_prec = self.prec();
        let r_assoc = self.r_assoc();

        op_prec > prec || r_assoc && op_prec == prec
    }

    fn prec(&self) -> Prec {
        match self {
            Self::LogicalOr => Prec::LogicalOr,
            Self::LogicalAnd => Prec::LogicalAnd,

            Self::Equal | Self::NotEqual => Prec::Equality,
            Self::Gt | Self::Lt | Self::GtEqual | Self::LtEqual => Prec::Comparison,

            Self::Add | Self::Sub => Prec::Term,
            Self::Mul | Self::Div | Self::Mod => Prec::Factor,
            Self::Pow => Prec::Exp,

            Self::Call => Prec::Call,
            Self::Access => Prec::Access,
        }
    }

    fn r_assoc(&self) -> bool {
        matches!(self, Self::Pow)
    }
}

impl<'a> Parser<'a> {
    pub fn parse_expr(&mut self) -> ParseResult<Expr> {
        self.parse_prec(Prec::Lowest)
    }

    fn parse_prec(&mut self, prec: Prec) -> ParseResult<Expr> {
        let mut expr = self.parse_lhs()?;

        while let Some(tt) = self.tokens.peek() {
            let Some(op) = self.get_op(tt) else {
                break;
            };

            if !op.should_parse_in_prec(prec) {
                break;
            }

            let tt = self.tokens.next().unwrap();

            expr = match op {
                BinOp::Call => {
                    let TokenTree::Group { tokens, .. } = tt else {
                        unreachable!();
                    };

                    let mut parser = self.parser_for_tokens(tokens.into_iter());

                    let args =
                        parser.parse_list("arguments", tkind!(punct Comma), Parser::parse_expr)?;

                    Expr::Call {
                        callee: Box::new(expr),
                        args,
                    }
                }

                BinOp::Access => {
                    let rhs = self.parse_spanned(|p| p.parse_prec(op.prec()));
                    let rhs_expr = rhs.inner?;

                    if let Expr::Var(Var::Simple(ident)) = rhs_expr {
                        Expr::Var(Var::Field {
                            expr: Box::new(expr),
                            field: ident,
                        })
                    } else {
                        // span is always valid - `parse_prec` always consumes a token
                        self.report(ParseError::InvalidAccessor(rhs.span.unwrap()));
                        Expr::Dummy
                    }
                }

                _ => {
                    let rhs = self.parse_prec(op.prec())?;
                    Expr::BinOp {
                        op,
                        lhs: Box::new(expr),
                        rhs: Box::new(rhs),
                    }
                }
            }
        }

        Ok(expr)
    }

    fn parse_lhs(&mut self) -> ParseResult<Expr> {
        match self.tokens.next() {
            Some(TokenTree::Token(token)) => match token.kind {
                TokenKind::Const(idx) => Ok(Expr::Const(idx)),
                TokenKind::Ident(ident) => Ok(Expr::Var(Var::Simple(Ident::Unresolved(ident)))),

                kind if kind == tkind!(kwd If) => self.parse_if_expr(),

                kind if kind == tkind!(punct Sub) => {
                    let expr = self.parse_prec(Prec::Unary)?;
                    Ok(Expr::UnaryOp {
                        op: UnaryOp::Neg,
                        expr: Box::new(expr),
                    })
                }

                kind if kind == tkind!(punct Bang) => {
                    let expr = self.parse_prec(Prec::Unary)?;
                    Ok(Expr::UnaryOp {
                        op: UnaryOp::Not,
                        expr: Box::new(expr),
                    })
                }

                _ => Err(ParseError::unexpected(
                    "an expression",
                    Some(TokenTree::Token(token)),
                )),
            },

            Some(TokenTree::Group {
                bracket_kind: BracketKind::Round,
                tokens,
                ..
            }) => {
                if tokens.is_empty() {
                    Ok(Expr::Unit)
                } else {
                    let mut parser = self.parser_for_tokens(tokens.into_iter());
                    let expr = parser.parse_expr()?;
                    parser.expect_end("parenthesised expression")?;
                    Ok(expr)
                }
            }

            other => Err(ParseError::unexpected("an expression", other)),
        }
    }

    fn get_op(&self, tt: &TokenTree) -> Option<BinOp> {
        Some(match tt {
            TokenTree::Token(t) => match t.kind {
                t if t == tkind!(kwd LogicalOr) => BinOp::LogicalOr,
                t if t == tkind!(kwd LogicalAnd) => BinOp::LogicalAnd,

                t if t == tkind!(punct EqualEqual) => BinOp::Equal,
                t if t == tkind!(punct BangEqual) => BinOp::NotEqual,

                t if t == tkind!(punct Gt) => BinOp::Gt,
                t if t == tkind!(punct Lt) => BinOp::Lt,
                t if t == tkind!(punct GtEqual) => BinOp::GtEqual,
                t if t == tkind!(punct LtEqual) => BinOp::LtEqual,

                t if t == tkind!(punct Add) => BinOp::Add,
                t if t == tkind!(punct Sub) => BinOp::Sub,

                t if t == tkind!(punct Mul) => BinOp::Mul,
                t if t == tkind!(punct Div) => BinOp::Div,
                t if t == tkind!(punct Mod) => BinOp::Mod,
                t if t == tkind!(punct Pow) => BinOp::Pow,

                t if t == tkind!(punct Dot) => BinOp::Access,

                _ => return None,
            },

            TokenTree::Group {
                bracket_kind: BracketKind::Round,
                ..
            } => BinOp::Call,

            _ => return None,
        })
    }

    fn parse_if_expr(&mut self) -> ParseResult<Expr> {
        let mut branches = vec![IfBranch {
            cond: self.parse_expr()?,
            then: self.parse_block()?,
        }];

        while self.eat_kind(tkind!(kwd Elif)) {
            let cond = self.parse_expr()?;
            let then = self.parse_block()?;
            branches.push(IfBranch { cond, then });
        }

        let else_ = if self.eat_kind(tkind!(kwd Else)) {
            Some(Box::new(self.parse_block()?))
        } else {
            None
        };

        Ok(Expr::If { branches, else_ })
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_ron_snapshot;

    use crate::parser::test_parse;

    #[test]
    fn l_assoc() {
        assert_ron_snapshot!(test_parse("1 + 2 + 3", |p| p.parse_expr()));
    }

    #[test]
    fn r_assoc() {
        assert_ron_snapshot!(test_parse("1 ^ 2 ^ 3", |p| p.parse_expr()));
    }

    #[test]
    fn mul_add() {
        assert_ron_snapshot!(test_parse("1 * 2 + 3", |p| p.parse_expr()));
    }

    #[test]
    fn add_mul() {
        assert_ron_snapshot!(test_parse("1 + 2 * 3", |p| p.parse_expr()));
    }

    #[test]
    fn mul_pow() {
        assert_ron_snapshot!(test_parse("1 * 2 ^ 3", |p| p.parse_expr()));
    }

    #[test]
    fn pow_mul() {
        assert_ron_snapshot!(test_parse("1 ^ 2 * 3", |p| p.parse_expr()));
    }

    #[test]
    fn group() {
        assert_ron_snapshot!(test_parse("(1)", |p| p.parse_expr()));
    }

    #[test]
    fn group_err() {
        assert_ron_snapshot!(test_parse("(1 1)", |p| p.parse_expr()));
    }

    #[test]
    fn unit() {
        assert_ron_snapshot!(test_parse("()", |p| p.parse_expr()));
    }

    #[test]
    fn if_expr() {
        assert_ron_snapshot!(test_parse("if a == b { 12 } else { 14 }", |p| p.parse_expr()));
    }

    #[test]
    fn elif() {
        assert_ron_snapshot!(
            test_parse("if a { 12 } elif b { 14 } else { 16 }", |p| p.parse_expr())
        );
    }

    #[test]
    fn trailing_else() {
        assert_ron_snapshot!(test_parse(
            "if a { 12 } elif b { 14 } else { 16 } else { 18 }",
            |p| p.parse_expr()
        ));
    }

    #[test]
    fn add_if() {
        assert_ron_snapshot!(test_parse("if a { 12 } else { 14 } + 2", |p| p.parse_expr()))
    }

    #[test]
    fn method_call() {
        assert_ron_snapshot!(test_parse("a.b().c", |p| p.parse_expr()))
    }

    #[test]
    fn invalid_accessor() {
        assert_ron_snapshot!(test_parse("a.12", |p| p.parse_expr()))
    }
}
