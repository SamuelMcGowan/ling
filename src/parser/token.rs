use super::span::Span;

#[derive(Debug, Clone, Copy)]
pub struct Token {
    pub span: Span,
    pub kind: TokenKind,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    Punct(Punct),
    Keyword(Keyword),
    Ident(Ustr),
    Error(Error),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Punct {
    LParen,
    RParen,
    LBrace,
    RBrace,
    LBracket,
    RBracket,

    Dot,
    Comma,
    Colon,
    Semicolon,

    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Pow,

    AddEqual,
    SubEqual,
    MulEqual,
    DivEqual,
    ModEqual,
    PowEqual,

    Gt,
    Lt,
    Equal,

    GtEqual,
    LtEqual,
    EqualEqual,

    FatArrow,
    Arrow,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Keyword {
    Func,
    Return,

    Type,
    Enum,
    Data,
    Typeclass,
    Impl,

    Let,
    Match,

    If,
    Elif,
    Else,

    Loop,
    While,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Error {
    UnexpectedChar(char),
}

macro_rules! tkind {
    (punct $punct:ident) => {{
        use crate::parser::token::*;
        TokenKind::Punct(Punct::$punct)
    }};
    (kwd $kwd:ident) => {{
        use crate::parser::token::*;
        TokenKind::Keyword(Keyword::$kwd)
    }};
    (ident $ident:expr) => {{
        use ustr::Ustr;
        use crate::parser::token::*;

        TokenKind::Ident(Ustr::from($ident))
    }};
    (error $error:ident $($value:tt)?) => {{
        use crate::parser::token::*;
        TokenKind::Error(Error::$error $($value)?)
    }}
}

pub(crate) use tkind;
use ustr::Ustr;
