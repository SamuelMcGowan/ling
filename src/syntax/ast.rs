use macro_rules_attribute::{derive, derive_alias};
use ustr::Ustr;

use crate::constants::ConstIdx;
use crate::symbol_table::SymbolId;

use super::parser::ParseResult;
use super::source::Span;

derive_alias! {
    #[derive(Node!)] = #[derive(
        Debug,
        Clone,
        serde::Serialize,
    )];
}

pub(crate) struct Spanned<T> {
    pub(crate) inner: T,
    pub(crate) span: Option<Span>,
}

impl<T> Spanned<ParseResult<T>> {
    pub fn transpose(self) -> ParseResult<Spanned<T>> {
        self.inner.map(|inner| Spanned {
            inner,
            span: self.span,
        })
    }
}

#[derive(Node!)]
pub(crate) struct Module {
    pub items: Vec<Item>,
}

#[derive(Node!)]
pub(crate) enum Item {
    Func(Func),
    Dummy,
}

#[derive(Node!)]
pub(crate) struct Func {
    pub ident: Ident,
    pub params: Vec<(Ident, Ty)>,
    pub ret_ty: Ty,
    pub body: Block,
}

#[derive(Node!)]
pub(crate) enum Ty {
    Constructed { ident: Ident, params: Vec<Ty> },
    Unit,
}

#[derive(Node!)]
pub(crate) struct Block {
    pub stmts: Vec<Stmt>,
    pub final_expr: Expr,
}

#[derive(Node!)]
pub(crate) enum Stmt {
    Expr(Expr),

    Loop(Block),
    WhileLoop { cond: Expr, block: Block },

    Declaration { lhs: Ident, rhs: Expr },
    Assignment { lhs: Var, rhs: Expr },

    Dummy,
}

impl Stmt {
    pub fn expect_delim(&self) -> bool {
        match self {
            Self::Expr(expr) => expr.expect_delim(),

            Self::Loop(_) | Self::WhileLoop { .. } => false,

            Self::Declaration { .. } => true,
            Self::Assignment { .. } => true,

            // the recovery function should have recovered past any semicolons
            Self::Dummy => false,
        }
    }
}

#[derive(Node!)]
pub(crate) enum Expr {
    Const(ConstIdx),

    UnaryOp {
        op: UnaryOp,
        expr: Box<Expr>,
    },

    BinOp {
        op: BinOp,
        lhs: Box<Expr>,
        rhs: Box<Expr>,
    },

    Call {
        callee: Box<Expr>,
        args: Vec<Expr>,
    },

    If {
        branches: Vec<IfBranch>,
        else_: Option<Box<Block>>,
    },

    Var(Var),

    Unit,

    Dummy,
}

impl Expr {
    pub fn expect_delim(&self) -> bool {
        !matches!(self, Self::If { .. })
    }
}

#[derive(Node!, Copy)]
pub enum UnaryOp {
    Neg,
    Not,
}

#[derive(Node!, Copy)]
pub enum BinOp {
    LogicalAnd,
    LogicalOr,

    Equal,
    NotEqual,
    Gt,
    Lt,
    GtEqual,
    LtEqual,

    Add,
    Sub,
    Mul,
    Div,
    Mod,

    Pow,

    Call,
    Access,
}

#[derive(Node!)]
pub(crate) struct IfBranch {
    pub(crate) cond: Expr,
    pub(crate) then: Block,
}

#[derive(Node!)]
pub(crate) enum Var {
    Simple(Ident),
    Field { expr: Box<Expr>, field: Ident },
}

#[derive(Node!)]
pub(crate) enum Ident {
    Unresolved(Ustr),
    Resolved(SymbolId),
}

impl Ident {
    pub fn unresolved(&self) -> Option<Ustr> {
        match self {
            Self::Unresolved(s) => Some(*s),
            Self::Resolved(_) => None,
        }
    }
}
