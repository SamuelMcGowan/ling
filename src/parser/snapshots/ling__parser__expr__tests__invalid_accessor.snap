---
source: src/parser/expr.rs
expression: "test_parse(\"a.12\", |p| p.parse_expr())"
---
(Ok(Dummy), "error: invalid accessor\n  ┌─ test_source:1:3\n  │\n1 │ a.12\n  │   ^^\n\n")
